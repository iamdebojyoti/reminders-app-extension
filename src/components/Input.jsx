import { Component } from "react";
import { Alert, Button, Col, Form, Row } from "react-bootstrap";
import PropTypes from "prop-types";

class Input extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "",
      description: "",
      added: false,
      valid: true,
      show: false,
    };
    this.save = this.save.bind(this);
  }

  save() {
    if (this.state.name === "" || this.state.description === "") {
      this.setState({ valid: false, show: true });
    } else {
      this.props.addReminder({
        name: this.state.name,
        description: this.state.description,
      });
      this.setState({
        name: "",
        description: "",
        added: false,
        valid: true,
        show: false,
      });
    }
  }

  render() {
    const isValid = this.state.valid;

    return (
      <>
        <Form>
          <Form.Group as={Row} className="mb-3" >
            <Col sm="12">
              <Form.Control
                id="reminderName"
                value={this.state.name}
                type="text"
                placeholder="Reminder Name"
                onChange={e => this.setState({ name: e.target.value })}
              />
            </Col>
          </Form.Group>

          <Form.Group as={Row} className="mb-3" >
            <Col sm="12">
              <Form.Control
                id="reminderDesc"
                value={this.state.description}
                as="textarea"
                placeholder="Description"
                onChange={e => this.setState({ description: e.target.value })}
              />
            </Col>
          </Form.Group>

          <Col sm="12">
            <Button type="button" onClick={this.save}>
              Submit
            </Button>
          </Col>
        </Form>

        {!isValid && (
          <Alert
            variant="danger"
            onClose={() => this.setState({ show: false })}
            dismissible
          >
            <Alert.Heading>Input Error</Alert.Heading>
            <p>Inputs cannot be empty</p>
          </Alert>
        )}
      </>
    );
  }
}

Input.propTypes = {
  addReminder: PropTypes.func.isRequired,
};

export default Input;
