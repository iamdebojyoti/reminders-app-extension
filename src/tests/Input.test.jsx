import { shallow } from "enzyme";
import Input from "../components/Input";

describe('Testing <Input/> component', () => {

    let component, props;

    beforeEach(() => {
        props = {
          addReminder: jest.fn(),
        };
    });

    it('should match snapshot of Input at initial state', () => {
      component = shallow(<Input {...props}/>);
      expect(component.debug()).toMatchSnapshot();
    });

    it('should match snapshot of Input when submit is clicked and no inputs are provided', () => {
        component = shallow(<Input {...props}/>);
        component.find('Button').simulate('click');
        component.update();
        expect(component.debug()).toMatchSnapshot();
    });

    it('should match snapshot of Input after submit is clicked when inputs are provided', () => {
        component = shallow(<Input {...props}/>);
        component.find('#reminderName').simulate('change', {target:{value:'Reminder 1'}});
        component.find('#reminderDesc').simulate('change', {target:{value:'Reminder 1 Desc'}});
        component.find('Button').simulate('click');
        component.update();
        expect(component.debug()).toMatchSnapshot();
    });

    it('should be able to add reminder when inputs are provided', () => {
        component = shallow(<Input {...props}/>);
        component.find('#reminderName').simulate('change', {target:{value:'Reminder 1'}});
        component.find('#reminderDesc').simulate('change', {target:{value:'Reminder 1 Desc'}});
        component.find('Button').simulate('click');
        component.update();
        expect(props.addReminder).toHaveBeenCalledWith({
            name: 'Reminder 1',
            description: 'Reminder 1 Desc'
        });
    });

    it('should not be able to add reminder when inputs are empty', () => {
        component = shallow(<Input {...props}/>);
        component.find('Button').simulate('click');
        component.update();
        expect(component.state('valid')).toBe(false);
    });
    
  });