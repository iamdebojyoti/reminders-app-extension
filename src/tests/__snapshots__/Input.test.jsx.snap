// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Input/> component should match snapshot of Input after submit is clicked when inputs are provided 1`] = `
"<Fragment>
  <Form>
    <FormGroup as={{...}} className=\\"mb-3\\">
      <Col sm=\\"12\\">
        <FormControl id=\\"reminderName\\" value=\\"\\" type=\\"text\\" placeholder=\\"Reminder Name\\" onChange={[Function: onChange]} />
      </Col>
    </FormGroup>
    <FormGroup as={{...}} className=\\"mb-3\\">
      <Col sm=\\"12\\">
        <FormControl id=\\"reminderDesc\\" value=\\"\\" as=\\"textarea\\" placeholder=\\"Description\\" onChange={[Function: onChange]} />
      </Col>
    </FormGroup>
    <Col sm=\\"12\\">
      <Button type=\\"button\\" onClick={[Function: bound save]} variant=\\"primary\\" active={false} disabled={false}>
        Submit
      </Button>
    </Col>
  </Form>
</Fragment>"
`;

exports[`Testing <Input/> component should match snapshot of Input at initial state 1`] = `
"<Fragment>
  <Form>
    <FormGroup as={{...}} className=\\"mb-3\\">
      <Col sm=\\"12\\">
        <FormControl id=\\"reminderName\\" value=\\"\\" type=\\"text\\" placeholder=\\"Reminder Name\\" onChange={[Function: onChange]} />
      </Col>
    </FormGroup>
    <FormGroup as={{...}} className=\\"mb-3\\">
      <Col sm=\\"12\\">
        <FormControl id=\\"reminderDesc\\" value=\\"\\" as=\\"textarea\\" placeholder=\\"Description\\" onChange={[Function: onChange]} />
      </Col>
    </FormGroup>
    <Col sm=\\"12\\">
      <Button type=\\"button\\" onClick={[Function: bound save]} variant=\\"primary\\" active={false} disabled={false}>
        Submit
      </Button>
    </Col>
  </Form>
</Fragment>"
`;

exports[`Testing <Input/> component should match snapshot of Input when submit is clicked and no inputs are provided 1`] = `
"<Fragment>
  <Form>
    <FormGroup as={{...}} className=\\"mb-3\\">
      <Col sm=\\"12\\">
        <FormControl id=\\"reminderName\\" value=\\"\\" type=\\"text\\" placeholder=\\"Reminder Name\\" onChange={[Function: onChange]} />
      </Col>
    </FormGroup>
    <FormGroup as={{...}} className=\\"mb-3\\">
      <Col sm=\\"12\\">
        <FormControl id=\\"reminderDesc\\" value=\\"\\" as=\\"textarea\\" placeholder=\\"Description\\" onChange={[Function: onChange]} />
      </Col>
    </FormGroup>
    <Col sm=\\"12\\">
      <Button type=\\"button\\" onClick={[Function: bound save]} variant=\\"primary\\" active={false} disabled={false}>
        Submit
      </Button>
    </Col>
  </Form>
  <Alert variant=\\"danger\\" onClose={[Function: onClose]} dismissible={true} show={true} transition={{...}} closeLabel=\\"Close alert\\">
    <AlertHeading>
      Input Error
    </AlertHeading>
    <p>
      Inputs cannot be empty
    </p>
  </Alert>
</Fragment>"
`;
